src/pygenstability/data_clustering.py

```diff
@@ -0,0 +1,286 @@
+from scipy.spatial.distance import pdist, squareform
+from scipy.sparse.csgraph import minimum_spanning_tree
+from scipy.sparse import csr_matrix
+
+from pygenstability.pygenstability import run as pgs_run
+from pygenstability.plotting import plot_scan as pgs_plot_scan
+from pygenstability.optimal_scales import identify_optimal_scales
+
+
+def compute_kNN(D, k=5):
+    """Computes kNN graph."""
+
+    N = D.shape[0]
+
+    # get k nearest neighbours for each point
+    k_neighbours = np.argsort(D, axis=1)[:, 1 : k + 1]
+
+    # initialise adjacency matrix
+    A = np.zeros((N, N))
+
+    # build kNN graph
+    for i in range(N):
+        for neighbour in k_neighbours[i]:
+            A[i, neighbour] = D[i, neighbour]
+            A[neighbour, i] = A[i, neighbour]
+
+    return A
+
+
+def compute_CkNN(D, k=5, delta=1):
+    """Computes CkNN graph."""
+    # obtain rescaled distance matrix, see CkNN paper
+    darray_n_nbrs = np.partition(D, k)[:, [k]]
+    ratio_matrix = D / np.sqrt(darray_n_nbrs.dot(darray_n_nbrs.T))
+    # threshold rescaled distance matrix by delta
+    A = D * (ratio_matrix < delta)
+    return A
+
+
+class GraphConstruction:
+    """Graph construction."""
+
+    def __init__(
+        self,
+        metric="euclidean",
+        method="cknn",
+        k=5,
+        delta=1.0,
+        distance_threshold=np.inf,
+    ):
+        # parameters
+        self.metric = metric
+        self.method = method
+        self.k = k
+        self.delta = delta
+        self.distance_threshold = distance_threshold
+
+        # attributes
+        self.adjacency_ = csr_matrix
+
+    def fit(self, X):
+        """Construct graph from samples-by-features matrix."""
+
+        # if precomputed take X as adjacency matrix
+        if self.method == "precomputed":
+            assert (
+                X.shape[0] == X.shape[1]
+            ), "Precomputed matrix should be a square matrix."
+            self.adjacency_ = X
+            return self.adjacency_
+
+        # compute normalised distance matrix
+        D = squareform(pdist(X, metric=self.metric))
+        D_norm = D / np.amax(D)
+
+        # compute normalised similarity matrix
+        S = 1 - D_norm
+
+        # sparsify distance matrix with CkNN or kNN method
+        if self.method == "cknn":
+            sparse = compute_CkNN(D_norm, self.k, self.delta)
+
+        elif self.method == "knn":
+            sparse = compute_kNN(D_norm, self.k)
+
+        # undirected distance backbone is given by sparse graph and MST
+        mst = minimum_spanning_tree(D_norm)
+        backbone = np.array((mst + mst.T + sparse + sparse.T) > 0, dtype=int)
+
+        # apply distance threshold to backbone
+        backbone *= np.array(D <= self.distance_threshold, dtype=int)
+
+        # adjacency matrix has weights of similarity matrix
+        self.adjacency_ = S * backbone
+
+        return self.adjacency_
+
+
+class DataClustering(GraphConstruction):
+    """Data clustering."""
+
+    def __init__(
```

Comment:
to simplify massively here, and make it more robust wrt to changes in main package, you can use something like `pg_kwargs` to pass all pgs args in one dict, then pass it to pgs_run below, so one can use only a few they want to modify from default, and if we change some in pgs, it will always work here. Also the code here will look much simpler!

Reply:
Also, it distinguises clearly what is for this workflow, and what is pgs parameters

Reply:
Passing the pgs args in a dictionary requires the users to write the dictionary themselves, which I think is quite inconvenient and also makes the user interface look worse in my opinion.

My idea is also that the DataClustering class serves as an alternative interface to pgs for people who prefer the sklearn like oop approach.

Reply:
no, you can use **kwargs
check many libraries like matplotlib, they always have this

Reply:
I fixed it accordingly and you were right that it simplifies things.